Hash-indexed object store (#3)
CasperLabs/ceps: text/0003-hash-indexed-object-store.md

# Hash-indexed object store

## Summary

[summary]: #summary

> RFC PR: [casperlabs/rfcs#0003](https://github.com/casperlabs/rfcs/pull/0003)

Reviewer: ```suggestion
CEP PR: [casperlabs/ceps#0003](https://github.com/casperlabs/ceps/pull/0003)
```

> There is a variety of "objects" floating around in our codebase at the moment: Deploys, WASM blobs, Blocks, DAG nodes and others. By standardizing them into a [Git](https://git-scm.com)-inspired object model we can simplify our networking, gossiping, retrieval, and storage logic greatly.

Reviewer: ```suggestion
There are a variety of "objects" defined in various parts of our code base (such as Deploys, WASM blobs, Blocks, and DAG nodes). This proposal recommends standardizing all such objects using a [Git](https://git-scm.com)-inspired object model with the intent of unifying and simplifying the platform's networking, gossiping, retrieval, check pointing, and storage logic.
```

## Motivation

[motivation]: #motivation

> One possible implementation of a mechanism for a new node joining the network is to provide it with a block and have it download all parents and dependencies up until the well-known genesis block.

Reviewer: ```suggestion
Nodes attempting to join an existing network must accumulate sufficient data to "catch up" to the network's current state. The current default model would require a new node to start from genesis and attempt to catch up by acquiring and executing each block in order; however this becomes increasingly problematic as the block height of the network increases. 

Another possible approach for a new node joining the network would be to provide the newly joining node with a more current block from a trusted source. The new node could run the chain backwards from that block, downloading all parents and dependencies until reaching the genesis block (which is trivially identifiable as it has no parent).
```

> With the current node architecture, this approach would require quite a few component instances: Separate instantiations of the gossiper and fetcher components for each type involved, which at the minimum are block and deploy if not counting DAG nodes. Storage needs to namespace according to types and component developers need to remember which effect to call for each. During all of this, the core logic remains the same for all IO operations of these different types though.

Reviewer: ```suggestion
With the current node architecture, this approach would require separate instances of the gossiper and fetcher components for each type involved (Block and Deploy at a minimum, perhaps also DAG nodes). Additionally the storage component would need to namespace according to each type and component developers would need to remember which effect to call for each type. However, the core logic would remain the same for all such operations.
```

> By unifying all these types, we can simplify all of our IO components drastically. Ensuring that all have a unique hash that includes a type-discriminating tag ensures that we can use the same namespace for identifiers.

Reviewer: ```suggestion
If we instead unified using the Git-inspired object model this CEP recommends, the logic would be drastically simplified, depending upon the combination of a unique hash and a data type tag to properly identify each chunk of data rather than type semantics.
```

> As a result, our IO layer will be greatly simplified, as well as its usage. Possible gains are easier snapshotting, DB handling and shipping data around in general. Speed-ups through external caching become very real (see [Future possibilities](#future-possibilities)).

Reviewer: ```suggestion
In addition to improved ease of use for component developers, this approach would also simplify snap shotting and data shipping, and lend itself more easily to external caching (see [Future possibilities](#future-possibilities)).
```

> Taking this one step further, we can express the tree structure given through hash references independent of the objects type as well, moving more IO logic away from business components.

Reviewer: > Taking this one step further, we can express the tree structure given through hash references independent of the objects type as well, moving more IO logic away from business components.

Recommend you either shore this argument up or strike it from the CEP.

> Taking this one step further, we can express the tree structure given through hash references independent of the objects type as well, moving more IO logic away from business components.

Author: > Highway's state unfortunately isn't really just a DAG anymore. We have dependencies like:
>
> Any endorsements of messages in set X, from at least ⅔ of the validators by weight.

I have no good suggestions for endorsements yet, I am currently thinking that they might be collected and attached to something in the consensus DAG, maybe even slapped (as a whole) onto a vote. They could also be considered equivalent in the same sense we consider evidence equivalent (see below).

Have a look at the [DAG example](https://github.com/marc-casperlabs/ceps/blob/hash-index-object-store/text/0003-hash-indexed-object-store.md#example-issues-with-the-consensus-dag), I believe it offers some solutions to this.

## Guide-level explanation

[guide-level-explanation]: #guide-level-explanation

(It is recommended to read the [prior art](#prior-art) section beforehand)

Every *object* that our node handles has common properties, namely:

* It can be serialized to binary storage.

> * It has a unique identifying hash that depends solely on its contents.

Reviewer: > The implication here is that it depends upon all of its contents...

@Fraser999 

I believe this can be made to be true in all cases via aggregation, with the identifying hash of an aggregate being the hash of the hashes of its sub-elements.

> * It has a unique identifying hash that depends solely on its contents.

Reviewer 2: For something like a `Deploy` where it needs to be signed by all parties before it's given to the network, then I can see this being workable.  But for `Block`s which get stored and have further signatures appended as time passes, it'd be pretty convoluted I think.

The signatures would need to be of something like the current block hash (effectively the hash of everything in the block except the signatures themselves), but as we accumulate them into the block, the key under which it'd be stored will keep changing.

I imagine we'd be better keeping the block static under its block hash, and treating the signatures as separate metadata.  Ideally the changing metadata would also be stored under a fixed key (e.g. `hash(block_hash + "metadata")`), but that doesn't really fit well into the storage model proposed here I think.

> * It has a unique identifying hash that depends solely on its contents.

Author: @Fraser999 Please have a look at the [Finality Signatures](https://github.com/marc-casperlabs/ceps/blob/hash-index-object-store/text/0003-hash-indexed-object-store.md#finality-signatures) part, in my opinion the scheme there should address all the issues mentioned here specifically for finality signatures.

* It may reference other objects.

> This is common to everything we handle, for example a block will reference a number of deploys, which in turn reference WASM blobs.

Reviewer: ```suggestion
This is common to everything we handle, for example a Block will reference a number of Deploys, which in turn reference WASM blobs.
```

### Serialization

> Every object can be serialized to be sent across the network, which also means it can be stored in storage. In general, any serialized object is prefixed with a short tag indicating its type, regardless of the chosen serialization format.

Reviewer: ```suggestion
Using the proposed model, every object recognized by the platform could trivially be serialized, prefixed with a type tag, and sent across the network. 
```

> This helps disamgiuate between different objects whose serialization may be the same under a given format, and can also be used to avoid conflicts with versioning, should an existing storage be run with a different node version.
> 
> The tag should be kept short, one byte should be plenty to cover all of our usecases (we will reserve one value for future expansion if necessary).

Reviewer: ```suggestion
A type tag helps disambiguate objects whose serialization may be the same under a given format, and can also be used to avoid conflicts across different node versions. One byte should be sufficient to cover all of our use cases (one value could be reserved for future expansion if necessary).
```

### Hashing

> Object hashes are always SHA512 hashes, as these are already in use by the execution engine. It is imperative that we standardize on one kind of hash for all objects.

Author: We are currently considering Blake3 in https://github.com/CasperLabs/ceps/pull/8.

I did assume hashes were SHA512, which was wrong, I will correct that. In general, the hash kind is rather arbitrary, as long as it is used consistently across the platform =).

> Object hashes are always SHA512 hashes, as these are already in use by the execution engine. It is imperative that we standardize on one kind of hash for all objects.

Author: Fixed in https://github.com/marc-casperlabs/ceps/commit/4c249eddfb73be4e57a9b0283a6160f49627abe7

> The hash value is derived solely from the serialized representation of the object, including its tag. This has the advantage that hashes can be checked without deserializing the object, even progressively while downloading it from a peer if necessary.

Author: See [above](https://github.com/CasperLabs/ceps/pull/3#discussion_r471294850) for a potential solution. I am resolving this only to keep the discussion in one place :)

It also allows storage and hash verification by other tools that are unaware of any of the business logic.

### Object references

Any properly typed object in memory provides a method for iterating over all of its *dependencies*. In graph terms this is an iterator over all outgoing edges.

> With this functionality, the logic of fetching dependent objects is moved out of the business logic components and into the fetcher. Even "validation" (verifying all references parts are available) happens at this point, this consumes the entire current logic of the block validator.

Reviewer: ```suggestion
With this functionality, the logic of fetching dependent objects is moved out of the business logic components and into the fetcher. Verifying all referenced dependencies are available is subsumed into the fetcher as well. 

_NOTE: this would render the current block validator component obsolete._
```

> Other components can decide if they want to fetch or fetch recursive, optionally specifying a peer that is required to provide the specific missing parts when asked.

Author: I do agree now. The CEP has been changed accordingly =)

> Other components can decide if they want to fetch or fetch recursive, optionally specifying a peer that is required to provide the specific missing parts when asked.

Author: I removed the "optionally" from the CEP, which solves this, I think. Additional avenues like external caches or asking random other peers for data is an optimization at that point, right?

Put into "pseudo-git" representation, this fictional format would look like this

```text
block
era 123
parent 2a691f09b09c62fb99fd1b52cfdd0ffecb68871a9af40ab48662b3c943a87a82
deploy a4ddaa0f887c31af66fa74d2bcabfde7504ce2d044315f2d9e0691b89a33f135
deploy f70a43aa0735e622f419e4456d4105a18b75ae194a7e97a73acbc4065a33d4a2
signature d042c9d14a1097344437735b84052ea997c42ba8d4e54de034be3d19d92b0a13
...
```

```text
deploy
gas-limit 0x32312
code 8257512c1a16fbc4ad536d1603d6c8fdfe9f05a1b3d537e0b17770e0de01a43b
...
```

Code refers to another object that contains the WASM code.

> ## Reference-level explanation
> 
> [reference-level-explanation]: #reference-level-explanation
> 
> First, we introduce a trait for object types:

Reviewer: ```suggestion
## Reference-level explanation

[reference-level-explanation]: #reference-level-explanation

_NOTE: this section assumes adoption of [casperlabs/ceps#0002](https://github.com/casperlabs/ceps/pull/0002)_
```

```rust
type ObjectHash = Sha512;
type ObjectError = bincode::Error;

/// Maximum object size is 32M.
const OBJECT_SIZE_LIMIT: usize = 32 * 1024 * 1024;

/// Set appropriate bincode opts.

> // Note: Depends on RFC-2

Reviewer: ```suggestion
```

Author: Fixed.

            //       and add tests for known binary representations.
            //       If varint encoding works correctly, this can be a
            //       u32 easily.
pub enum Object {
    Block(Block),
    Deploy(Deloy),
    // ...
}

impl Object {
    /// Deserialize an object, calculating hash while doing so.
    fn deserialize_from<R: Read>(reader: R) -> Result<(Self, ObjectHash), ObjectError> {
        let opts = bincode_opts();
        let mut hr = HashReader512::new(reader);
        let value = opts.deserialize_from(&mut hr)?;

        let (_innerReader, hash) = hr.take();

        Ok((value, hash))
    }

    /// Serialize an object, calculation hash while doing so.
    fn serialize_into<W: Write>(&self, writer: W) -> Result<ObjectHash, ObjectError> {
        let opts = bincode_opts();
        let mut hw = HashWriter512::new(writer);

        opts.serialize_into(hw, &self)?;

        let (_innerWriter, hash) = hw.take();

        Ok(hash)
    }

    /// Calculate hash only. This function may be unnecessary.
    #[inline(always)]
    fn calc_hash(&self) -> Result<ObjectHash, ObjectError> {
        self.serialize(&mut NullWriter)
    }
}

> 
> /// An object that supports iteration over its dependencies.

Reviewer: ```suggestion
_NOTE: the custom iterator made necessary by the design of `iter_outgoing` does not require heap allocations._

/// An object that supports iteration over its dependencies.
```

Author: 2368116

pub trait OutIterable {
    type OutIter : Iterator<Item=ObjectHash>;

    /// Iterate over all hashes referenced by this object directly.
    fn iter_outgoing(self) -> Self::OutIter;
}

enum ObjectOutIter<'a> {
    Block(<Block as OutIterable>::OutIter<'a>),
    Deploy(<Deploy as OutIterable>::OutIter<'a>),
    // ...
}

impl<'a> OutIterable for &'a Object {
    type OutIter = ObjectOutIter<'a>;

    fn iter_outgoing(self) -> Self::OutIter {
        match self {
            Object::Block(block) => ObjectOutIter::Block(block.iter_outgoing()),
            Object::Deploy(deloy) => ObjectOutIter::Deploy(deploy.iter_outgoing()),
            // ...
        }
    }
}

impl Iterator for ObjectOutIter {
    type Item = ObjectHash;

    // ...
}
```

This implementation will likely require the implementation of three utility types, a `NullWriter` that just dicards data, a `HashWriter` and a `HashReader` that pass input through but update an internal hash each time `write` is called.

> Note that the hash type is hardcoded to keep things simple for now.
> 
> Let's implement this for the provisionary `Block` type. Note that the custom iterator is required by our design of `iter_outgoing`, but it does not require any heap allocations:

Reviewer: ```suggestion
_NOTE: hash type is hard coded to simplify the example._

EXAMPLE: `Block` type implementation. 
```

Author: 794a254

```rust
pub struct BlockHeader {
    parent_hash: ObjectHash,
    root_state_hash: ObjectHash,

>     consensus_data: Vec<ObjectHash>,

Author: This is meant to be example code, but I changed it to deploys for now.

    era: u64,
    proofs: Vec<Signature>,
}

#[derive(Debug, Deserialize, Serialize)]
enum BlockHeaderIter<'a> {
    Parent(&'a Block),
    RootState(&'a Block),
    ConsensusData(&'a Block, usize),
    Done,
}

impl<'a> OutIterable for &'a BlockHeader {
    type OutIter = BlockHeaderIter<'a>;

    fn iter_outgoing(self) -> Self::OutIter {
        BlockHeaderIter::Parent(self)
    }
}

> impl<'a> Iterator for IterState<'a> {

Author: Fixed in 7163333

    type Item = ObjectHash;

    fn next(&mut self) -> Option<Self::Item> {
        let mut rv = None;

        *self = match *self {
            BlockHeaderIter::Parent(block) => {
                rv = block.parent_hash.clone();
                IterState::RootState(block)
            }
            BlockHeaderIter::RootState(block) => {
                rv = block.root_state.clone();
                IterState::ConsensusData(block, 0)
            }
            BlockHeaderIter::ConsensusData(block, idx) if idx < block.consensus_data.len() => {
                rv = block.consensus_data[idx].clone();
                IterState::ConsensusData(block, idx+1)
            }
            BlockHeaderIter::ConsensusData(block, idx)
            | BlockHeaderIter::Done => BlockHeaderIter::Done,
        };

        rv
    }
}
```

> This is all we need to be able to store and send arbitrary objects. We add convenience methods for our known types here:

Reviewer: ```suggestion
_NOTE: The above logic is necessary to store and send arbitrary objects using this proposed model; the following logic describes convenience methods to improve the ergonomics for component developers._
```

```rust
impl Object {
    fn as_block(&self) -> Option<&Block> {
        match self {
            Object::Block(ref block) => Some(block),
            _ => None
        }
    }

>     fn as_block_mut(&mut self) -> Option<&mut Block> {

Author: I'll remove it and put in something regarding finality signatures.

        match self {
            Object::Block(ref mut block) => Some(block),
            _ => None
        }
    }

    // ...
}
```

### Storage

> The storage component (there is only a single one now) turns into an `ObjectStore`, with only a few effects left:

Reviewer: ```suggestion
Using this proposed model, the storage component decomplects into a single `ObjectStore`, with a simplified set of effects.
```

```rust
/// Retrieve multiple objects.
// This method always uses the heap, since a single is already quite large.
// Does not return any dependencies, may return less objects than requested.

> pub async fn get_objects(self, object_hash: Vec<ObjectHash>) -> HashMap<ObjectHash, Object>;

Author: If this is about the random order of `HashMap`, I am thinking that that is an entirely orthogonal problem to this CEP. For one, we should be in control of our tests random hashes, although the implementation details might be annoying (changing out the hasher). Second, I don't even mind using a `BTreeMap` if this is the pragmatic thing.

If we just specify a well defined order for all operations, we won't even have these issues - there's no rule that says we have to rely on the random order inside a hash map for anything. We could make networking sort all dependencies by hash before sending them, greatly reducing this issue.

/// Retrieve multiple objects, including their dependencies.
///
/// If specified, stops at a specific depth. Returns `None` if any requested
/// object was incomplete until `max_depth`.
pub async fn get_objects_with_dependencies(self, object_hash: Vec<ObjectHash>, max_depth: Option<NonZeroUsize>) -> Option<HashMap<ObjectHash, Object>>>;

/// Store multiple objects.
///
/// Objects are stored in the reverse order given, so dependencies should be
/// listed behind the actual object. Always recalculates object hashes for

> /// consistency reasons.

Author: "Invalid" objects, i.e. those where the hash does not match the object, should already be filtered at the boundary. We trust everything from our own store (although we might hash it "on the way up" when loading it for durability reasons). Incoming messages containing objects have their hash verified as the first thing that happens.

The only reason hashes are returned is that I had hoped to save us a redundant hashing step. Figuring out the optimal representation of objects for efficiency is non-trivial here. As an example, this is valid alternative:

```rust
struct ImmutableObject {
    /// Cache of the object hash.
    hash: ObjectHash,
    /// The actual object.
    inner: Object,
}
```

There are other ways with interior mutability that calculate these on the fly and there is always a trade-off.

The one common factor is we check objects as soon as possible, going in. I'll add the `ImmutableObject` to the CEP to make this a bit more obvious.

pub async fn put_objects(self, objects: Vec<Object>) -> Vec<ObjectHash>;
```

> The storage still uses LMDB as a backend, but calculates keys upfront (the implementor may consider writing to a temporary key and renaming it though) and stores the object under it. For convenience and since it is cheap, given a proper `Hash` trait implementation for `ObjectHash`, we already return hashmaps as many consumer are likely to use the results this way.

Reviewer: ```suggestion
The `ObjectStore` continues to use LMDB as a backend, but calculates keys upfront (but we may consider writing to a temporary key and renaming it) and stores the object under the precalculated key. For convenience, given a proper `Hash` trait implementation for `ObjectHash`, a hashmap might be returned.
```

> The recursive `_with_dependencies` retrieval functions allows fetching a whole subtree, which is useful when getting a block from storage - all the deploys (`max_depth` >= 1) and their WASM (`max_depth` is >= 2) can be fetched alongside with it. This saves a lot of event roundtrips.

Reviewer: ```suggestion
The recursive `_with_dependencies` retrieval functions allows fetching an entire subtree, which is useful when getting a block from storage. For instance, all contained deploys (`max_depth` >= 1) and their WASM (`max_depth` is >= 2) could be fetched alongside of a block, saving event round trips.
```

### Execution engine

The execution engine is going to require a changes, namely in the hashing function. All values it executes will need to be wrapped in an object enum before being stored/hashed. The exact impact of this is unknown, see the section on [unresolved questions](#unresolved-questions).

Additionally, if WASM blocks are to be split off, these will need to be fetched separately. If any of these solution provide to be infeasible, passing around a wrapper around LMDB like we do now is still feasible to keep these operations speedy.

### Gossiping

> The gossiper undergoes a few changes starting with no more generics. Existing functionality can be kept and the component can conditionally announce objects that are incoming deploys still.

Reviewer: ```suggestion
Under this proposed model, we no longer require a generic gossiper component; a single statically typed gossiper can handle all objects. The existing functionality can be kept, with some allowed specialization around announcements (such as an incoming deploy received announcement). 
```

> The gossiper supports an effect in the form of

Reviewer: ```suggestion
The proposed universal gossiper component would need to support a `gossip_objects` effect as described below.
```

```rust
/// Gossip a set of objects.
///
/// Dependencies of objects will not be retrieved, checked or gossiped,
/// it is up to the caller to ensure the object itself and its dependencies
/// are in the local storage.
pub async fn gossip_objects(self, objects: Vec<Object>);
```

> The algorithm works otherwise similar to how it works now.

Reviewer: ```suggestion
```

### Fetching

> The fetcher component also loses its type parameter, similar to the gossiper.

Reviewer: ```suggestion
Under this proposed model, we no longer require a generic fetcher component; a single statically typed fetcher can handle all objects. The existing functionality can be retained with some minor alterations such as the `fetch_object` function described below . 
```

```rust
/// Retrieve an object from storage or the network.
///
/// Retrieves an object
pub async fn fetch_object(self, object_hash: Object,
                                max_depth: Option<NonZeroUsize>,
                                peer: I) -> HashMap<ObjectHash, <FetchResult>>;
```

Here we employ the same `max_depth` limiter as we do with the storage itself.

> It is worth noting that given a current block hash `h`, calling `fetch_object(h, None, some_peer)` will eventually fetch the entire state required to join the network!

Author: > Is the global state considered a dependency? (Then this call would probably download not only the current, but also all historical states?)

Yes. But there is nothing stopping us from defining smarter graph traversal in the future that stops at an earlier point, or only downloading global states that we actually need. We have to decide to which extend nodes are "on the hook" for providing older global states. If pressed, they could always recalculate them though.

>  Or do we compute the state from genesis once we've downloaded the chain?

We could, that is up to us.

### DAG/Consensus

> There are two ways of embedding consensus-related DAG data into the system, either the object format accompanies every single type of supported consensus with an enum variant, or a generic variant is used.

Author: We will have to backwards-support all data though and do not want to reuse tags, right? So I am arguing that the enum variant is safer.

The first alternative is not as a bad as it sounds; while it does constitute a leaky abstraction, it improves the general robustness of the system, which will know how to reject invalid node types and can handle them "properly" inside its storage.

The second alternative is adding a generic

```rust
enum Object {
    // ...
    DagNode<Box<dyn DagNode>>
}
```

variant and implementing variations of these inside the consensus component. This will result in a trait and the consensus level having to do a bit of serialization/deserialization, but does not leak the associated discriminator of the enum into the serialized representation.

## Drawbacks

[drawbacks]: #drawbacks

> The biggest issue is that changes might affect execution engine performance negatively. This is still open to discovery, see [unresolved questions](#unresolved-questions).

Reviewer: ```suggestion
The proposed changes might negatively affect the contract runtime component's performance. This is still open to discovery, see [unresolved questions](#unresolved-questions).
```

> There is a considerable effort required to update all components and the change might not be easy to implement incrementally.

Reviewer: ```suggestion
This proposal represents a sweeping, cross cutting, and not easily iterable change affecting many components. 
```

> The whole change to the execution engine may be prohibitive because of external dependencies, i.e. consumers of said execution engine that are not our node. This _might_ be worked around with a trait that abstracts the coupling between `Object` and components away.

Reviewer: > The whole change to the execution engine may be prohibitive because of external dependencies, i.e. consumers of said execution engine that are not our node. This _might_ be worked around with a trait that abstracts the coupling between `Object` and components away.

It isn't apparent to me that this proposal actually affects global state at all; as far as I can tell the TrieStore would remain untouched. Please clarify?

Author: Ideally `TrieStore` store elements would also find their place in the unified object store, but this needs to be examined still =(

## Rationale and alternatives

[rationale-and-alternatives]: #rationale-and-alternatives

> ### Hashing without serialization

Reviewer: I would vote against this alternative, if we were even seriously considering it.

Author: Noted!

One alternative is to not hash serialized contents of objects, but have every object provide hashing functionality itself. This has the advantage of being a little cheaper should an object in memory need to be hashed without being serialized; however this is likely a rare occurence. It also makes the objects independent of the serialization method chosen and might reduce changes needed inside the execution engine.

> The drawback is that hashing now has to be implemented for each object on its own, but more importantly it is possible to accidentally introduce security issues by forgetting to include a field in a hash after changing it. Any occurence of such a bug is likely to result in a security issue, as it allows crafting two different specimen of an object with the same hash.

Author: a7e3fe8

This also removes easy verification of objects by outside sources.

### Keeping components generic

> We can keep the fetcher and gossiper generic to reuse more of the existing code. Under this model we will reduce the number of gossipers to two (e.g. network and objects) or three (network, objects, consensus). Moving the logic for "validating" (checking for completeness) of an object is also not necessarily moved into fetcher this way.

Reviewer: Retaining the generic versions and simply providing concrete impls for Object might shave a little time off the implementation. Seems pragmatic to me.

Author: I suspect (hope) the changes there would be small either way.

```console
$ git gc
...
$ find .git/objects/ -type f | wc -l
4
```

Where did our files go? They are in a pack file:

```console
$ git verify-pack --verbose ./.git/objects/pack/pack-80a4d68ade4dcac5913a0e825ff50c6cb968c5a8.idx
23d9244c918044b4f1e28de295065b41e029f37c commit 1122 848 12
ff51a7fc40b7fb6f165d3a92ab88d8d2bf8c0b18 commit 1117 844 860
...
```

Pack files are archives that contain multiple objects. They make things a little easier on the filesystem and the HTTP transfers, but more importantly delta compression is applied when files are in a pack.

### Git summary

Git stores different types of objects indexed by hashes, compressed in a very simple key-value scheme on the local filesystem. Every object contains a header indicating its type (commit, tree, blob). These objects can be fetched by hash even across the network. Parent relationships are indicated by hash as well.

## Unresolved questions

[unresolved-questions]: #unresolved-questions

* What is the actual impact on the execution engine?
* Do we serialize hashes or contents?
* Do we store DAG nodes statically or dynamically (`Box<dyn ...>`)?

> All these questions should be answered before this RFC is merged.

Reviewer: > All these questions should be answered before this CEP is merged.

Point of order question: do you want us to discuss these items on this CEP or do you want to plan one or more discovery tasks?

Author: We will likely need more of these, one think I can think of is breaking out a separate CEP that examines how the consensus DAG (as opposed to the chain DAG) can be made to fit this scheme (and whether or not it should).

But I consider general support for the approach laid out here a prerequisite.

> All these questions should be answered before this RFC is merged.

Reviewer: ```suggestion
All these questions should be answered before this CEP is merged.
```

Author: 335cc5c

## Future possibilities

[future-possibilities]: #future-possibilities

> While not covered in detail in this RFC, these changes lay the groundwork for exciting future possiblities:

Reviewer: ```suggestion
While not covered in detail in this CEP, this proposal also lays the groundwork for future possibilities:
```

Author: f6e37f3

### CDNs

> Since retrieving objects is just a matter of fetching by hash, we can leverage a CDN, for example via HTTP. Whenever a node needs an object `435abcf`, it can try to `GET http://casper.cdn.net/objects/435abcf` first. If found, we just saved the network the burden of sending the missing piece. This is not limited to the "official" CDN, validators are free to run one themselves, we may even add mechanisms to discover them.

Reviewer: ```suggestion
As under this proposal, retrieving objects is simply a matter of fetching by hash, it would be possible to trivially leverage a CDN to shift a significant amount of data servicing load off of operating nodes. 

For instance, whenever a node needs an object `435abcf`, if a HTTP based CDN were available that node could attempt to `GET http://casper.cdn.net/objects/435abcf` first before asking other nodes on the network to provide that data. This would not require a single centralized CDN either, validators or other parties could run their own CDN. Going a step further, we might even add a mechanism to discover available CDNs.
```

> Open questions with this approach is who puts data on the CDN in the first place:

Reviewer: ```suggestion
An obvious question with this approach would be, who puts data on a CDN in the first place?
```

> * An easy first take is to upload daily snapshots of the whole chain ourselves. This already considerably reduces the burden on the network when new nodes are joining.
> * Another approach is to allow any node to push objects to the CDN. This might require a plausibility check (i.e. show that the object is reachable from genesis) and thus specialized software.

Reviewer: ```suggestion
* An easy first implementation would be an opt-in (via config.toml) behavior for a node instance to upload daily snapshots of the entire chain to a target CDN, and an opt-in behavior to use one or more CDN's. CasperLabs operated nodes would operate a CDN and opt-in to both options using our own CDN. Node operators who trust use could opt-in to use the CDN we provide. Those who do not are free to wire up their own CDN if they like. 
** This minimal version would considerably reduce the burden on the network of new nodes joining, at the cost of periodic activity spikes on nodes pushing snap shots to a CDN.
* Another approach is to allow any node to push objects to a CDN. This might require a plausibility check (i.e. show that the object is reachable from genesis) and thus would need some specialized software to enable.
```

Author: Manually added in 463121f

> * Clients could also upload WASM blocks to a CDN beforehand, making the actual deploy really small.

Reviewer: ```suggestion
* Clients could also upload WASM blocks to a CDN beforehand, significantly reducing the size of a referencing Deploy.
```

Author: daf1412

> Note that none of this is strictly required as all the CDN functionality is strictly optional with regards to the correct operation of the network.

Reviewer: ```suggestion
_NOTE: the described CDN functionality would be strictly optional with regards to the correct operation of the network._
```

Author: 4e99351

### Git-style pack files

> We may copy the idea of pack files, saving us some roundtrips. This is the first more complicated feature, but a node joining could easily happen to find a pack file that contains the
> blocks (without deploy contents) all the way up to genesis. It will then fetch other information as required.

Reviewer: ```suggestion
We could opt to implement pack files, to further optimize and most notably reduce total round trips. This is a non-trivial optimization, but offers significant benefits. For instance, it is easy to imagine a new node receiving a pack file that contains all Blocks (without dependencies) all the way back to genesis; that node could then fetch other information as required.
```

Author: 4e99351

> A pack is just a collection of objects, so importing one into a node should be (disk space not withstanding) strictly a benefit to the node. In the absence of CDNs, pack files can be used to speed up inter-node downloads.

Reviewer: ```suggestion
A pack is just a collection of objects, so importing one into a node should be (disk space not withstanding) strictly a benefit to the node. This idea is also somewhat orthogonal to CDN usage, as even if we did not opt to support CDNs a pack file format could still be used to speed up inter-node downloads.
```

Author: 21c6cb6

> It is entirely possible to implement pack files as just a serialization of `HashMap<ObjectHash, Object>`.

Author: Yes, Pack files solve the "many small files" issues on the network, not locally. We want to avoid making a lot of requests though, instead sending small bits in bulk.

It also allows for easy shipping of archives.

### Snapshots

A feature currently on the roadmap becomes almost a non-issue with this approach; snapshotting or *checkpointing*, i.e. the ability to start a node without having to download all content from the network, requires no code modification in the sense that if objects are identified by immutable hashes, the local database can simply check if they are present to avoid redownloading them.

> Combined with packfiles, a snapshot or *checkpoint* of a node is just a specific block with all its dependencies in a single block file.

Reviewer: ```suggestion
Combined with pack files, a snapshot or *checkpoint* of a node is just a specific Block plus all of its dependencies in a single file.
```

### WASM deduplication

> A good thing to factor out from deploys is the WASM code blob, this may mean to do away with deploy headers entirely if the biggest part of the deploy is now given by reference instead of by value. With some changes to the deploy format or smart contract tooling, it may be possible to reference multiple WASM blobs in a single deploy; in a good design, an installed WASM blob would be found under the same hash in both the deploy and the global state.

Reviewer: ```suggestion
A good thing to factor out from Deploys is the WASM code blobs. We might also do away with DeployHeaders entirely if the biggest part of the Deploy is now given by reference instead of by value. 

With some changes to the deploy format or smart contract tooling, it may be possible to reference multiple WASM blobs in a single deploy; in a good design, an installed WASM blob would be found under the same hash in both the deploy and the global state.
```

> ### Offloading
> 
> External storage for a node could be added, e.g. if node operators want to reduce disk pressure, the node could support offloading of objects to a store like [S3](https://aws.amazon.com/s3/), [Google Storage](https://cloud.google.com/storage) or [minio](https://min.io/). This essentially gives the node operators heaps of possibilties to scale out their infrastructure, without tying the chain to any of these providers --- these are all optional optimizations.

Reviewer: ```suggestion
### Offloading / Archiving

Support for external storage of older data could be added to allow node operators to reduce disk pressure. Offloading of objects to a store like [S3](https://aws.amazon.com/s3/), [Google Storage](https://cloud.google.com/storage) or [minio](https://min.io/) would offer node operators a significant (but optional) tool for scaling their infrastructure, without tying the chain itself to any specific service provider.
```

### Storing post state hashes

> Post state hashes refer to global state, which already forms a tree structure. If the blocks were to store a post state hash inside each block, a client could download only the chain until genesis as additional verification, but then know that the post state hash is correct.

Author: This is assuming they have a "trusted root", I will update the text accordingly to make this more clear.

Fetching the associated global state again just becomes a recursive (ideally CDN-powered, at least pack supported) download of the remaining hashes. This willl result in automatic "pruning", since outdated/overwritten data is not retrieved.

This is analogous to a "shallow clone" of a Git repository.

> As a whole, this topic is fairly complex and definitely warrants its own RFC.

Reviewer: ```suggestion
However, that is a fairly complex topic and beyond the scope of this CEP.
```

### Garbage collection

> Over time, the node will accumulate some dead weight, especially if the DAG is also stored inside the object store. A garbage collector can be run periodically to remove these keys from the object store backend. This process is also similar to Git, which considers any object pointed by a ref (like `HEAD`, branches or tags) worth keeping and eliminates those objects that are unreachable from any of these *tips*.

Reviewer: ```suggestion
An operating node will accumulate no longer used data over time, especially if DAG state is periodically persisted. A garbage collector should be run periodically to remove unreachable tips, allowing disk recovery. 
```

When designing this feature, there may be a chance to build in a "shallow" node mode where only the latest global state and minimal history is kept to reduce disk space.

For performance reasons, garbage collection may be run online, that is during normal operation. It should be feasible to keep a cache of all reachable object hashes in memory and keep it up-to-date seperately, especially given our serialized access due to the reactor architecture. With this in place, once the list is completey and guaranteed to be continuously kept up-to-date, the garbage collector can start walking all keys and eliminate unreachable ones.

### In-memory deduplication

Storage could keep a cache of recently used objects and use `Rc`s or `Arcs` to pass these around. This is usefuly because it is likely that a recently received deploy will make its way into a block in the near future. Having an unified mechanism that works for any kind of object makes implementing this case a good investment, as it will immediately benefit the whole system.

### Persisting networking state

> Should the networking component to be reworked to gossip data for node discovery, this data will need to be stored as well. It is conceivable that this might be stored to speed up a rejoin-process. This does, however, require a bit of indexing on the storage component's part, as there is currently no way to find all items of a specific object type.

Reviewer: ```suggestion
If the networking component is reworked to gossip data for node discovery, this data will need to be stored as well. It is conceivable that this might be stored to speed up a rejoin-process. This does, however, require a bit of indexing on the storage component's part, as there is currently no way to find all items of a specific object type.
```

## Motivation

[motivation]: #motivation

Nodes attempting to join an existing network with knowledge of a recent, trusted block hash must accumulate sufficient data to "catch up" to the network's current state. The current default model would require a new node to start from genesis and attempt to catch up by acquiring and executing each block in order; however this becomes increasingly problematic as the block height of the network increases.

With the current node architecture, this approach would require separate instances of the gossiper and fetcher components for each type involved (Block and Deploy at a minimum, perhaps also DAG nodes). Additionally the storage component would need to namespace according to each type and component developers would need to remember which effect to call for each type. However, the core logic would remain the same for all such operations.

> With the current node architecture, this approach would require quite a few component instances: Separate instantiations of the gossiper and fetcher components for each type involved, which at the minimum are block and deploy if not counting DAG nodes. Storage needs to namespace according to types and component developers need to remember which effect to call for each. During all of this, the core logic remains the same for all IO operations of these different types though.

Author: Fixed.

## Guide-level explanation

[guide-level-explanation]: #guide-level-explanation

(It is recommended to read the [prior art](#prior-art) section beforehand)

Every *object* that our node handles has common properties, namely:

* It can be serialized to binary storage.
* It has a unique identifying hash that depends solely on its contents.
For example a Block may reference a number of Deploys by hash, which in turn may reference Wasm blobs by hash.

This is common to everything we handle, for example a block will reference a number of deploys, which in turn reference Wasm blobs.

### Serialization

Using the proposed model, every object recognized by the platform could trivially be serialized, prefixed with a type tag, and sent across the network.

A type tag helps disambiguate objects whose serialization may be the same under a given format, and can also be used to avoid conflicts across different node versions. One byte should be sufficient to cover all of our use cases (one value could be reserved for future expansion if necessary).

> The tag should be kept short, one byte should be plenty to cover all of our usecases (we will reserve one value for future expansion if necessary).

Author: Yeah, must have slipped through. There is (was?) a bug where GitHub replaced the wrong line when clicking "Accept suggestion" in the review...

Below is a **simplified** example, in which we look at a fictional blockchain with three blocks (Gensesis, Block1, Block2), two of which containing two Deploys each. Some of these Deploys refer to Wasm-Blocks, one of which is shared. Block1 also happens to be a switch block, containing a list of validators for era 5:

![Simple example graph](../diagrams/0008-example.png)

Here the hash of Block2 (`a1b2c3d`) is enough to retrieve Block2, from which all other data can be reached. The Wasm1 blob that is shared between Deploy2 and Deploy3 is what prevents this graph from being a tree.

_NOTE: This example does not show the consensus DAG or global state root hash, see below for more information_

### Serialization

Every object recognized by the platform can trivially be serialized, prefixed with a type tag, and sent across the network or stored. The tag should be kept short, one byte should be plenty to cover all of our usecases (we will reserve one value for future expansion if necessary).

### Hashing

> Object hashes are always Blake2b256 hashes, as these are already in use by the execution engine. It is imperative that we standardize on one kind of hash for all objects, even though the actual hash algorithm is less important (see also: <https://github.com/CasperLabs/ceps/pull/8>).

Author: This is a weak rebuttal, but I would think that changing hashes later on is not as easy as change a type in the impl, especially when there's data already out there. Personally I am willing to byte one of these bullets usually and try compensate for the performance loss some other way, but I can't offer a silver bullet here, unfortunately.

The hash value is derived solely from the serialized representation of the object, including its tag. This has the advantage that hashes can be checked without deserializing the object, even progressively while downloading it from a peer if necessary.

It also allows storage and hash verification by other tools that are unaware of any of the business logic.

### Object references

Any properly typed object in memory provides a method for iterating over all of its *dependencies*. In graph terms this is an iterator over all outgoing edges.

With this functionality, the logic of fetching dependent objects is moved out of the business logic components and into the fetcher. Verifying all referenced dependencies are available is subsumed into the fetcher as well.

> _NOTE: this would render the current block validator component obsolete._

Author: It's certainly not the goal of this CEP, just a side-effect ;)

The consensus DAG is structure parallel to the chain DAG, while there may be "soft" references between the two, they are not made hard dependencies to avoid having to preserve the full consensus DAG for a specific era forever

## Refs

In this example given earlier in this section, a specific hash was designated as the starting point (`a1b2c3d` in the figure). Similar to Git references (tags/branches), we introduce these *refs*, which map a name to a hash.

A good example is for a ref is `refs/trusted`, which refers to the latest block that the node considers as finalized and valid, i.e. trustworthy. Should the node crash, it will resume operation by ensuring it has all the dependencies of `refs/trusted` and checking whether it is still inside the unbonding period (further details on node rejoining are out of scope for this CEP, see [CEP5](https://github.com/CasperLabs/ceps/pull/5) instead).

Building on this idea, `refs/latest` can refer to the latest block created known by the node, which may be newer than `refs/trusted`.

> The immediate gain is that different kinds of information retrieval can be standardized as well by asking other nodes for refs instead of adding custom network message. For example, should a node want to catch up with the network, it requests `refs/trusted` from another node through a generic "retrieve ref" interface as opposed to a specialized "retrieve latest block hash".

Author: If you debug the messages, you will see things like `GetRef("refs/trusted")`, which is arguably barely less easy to understand and debug than `GetTrustedHash`, I'd say.

In addition, it also stores references under their name, prefixed with `refs/`, as the prefix can never be part of a hex-formatted hash, thus automatically namespacing objects and refs.

## Type safety

An often brought up concern is the perceived type-safety loss by having a generic *object* as the central focus of this CEP. However, this is addressed by the following factors:

* Every serialized object carries a type annotation in its hashed serialization, thus the kind of an object is always known at runtime.
* Objects can be subjected to checked downcasts, e.g. an `as_block(&self) -> Option<&Block>` method that will only return a `Some` if the object in question is indeed a block.
* The API for retrieving can conveniently allow specificying a type and roll up downcast errors into the retrieval error type. As an example, a method `get_object_from_storage<T>(&self) -> Result<T, Error>` can be called with `T = Object` for no downcasts or a `T = Block` if a block is expect, combining "not found" and "wrong errors" in `Error`.
* Objects can still be passed around in their downcast form, only being upcast at IO barriers like storage or networking.

> Overall the amount of type safety retained or even added is firmly in the hand of the implementor of this CEP.

Author: There's nothing stopping us from wrapping these in newtypes though, should this be an issue or having the "Hash" part be a trait. The type safety can be layered on top of this, without any runtime overhead!

This, of course, is not entirely accurate - by attaching a hash of the global state's *root hash* (named *post state hash* when the hash is the result of the execution of a set of deploys from the current block), this process can be cut short:

![Diagram depicting the relationship between previous post state hashes, blocks and deploys](../diagrams/0008-global-state.png)

The proposed unified object model makes it trivial to download the global state by simply asking another node (or external caches, see [Future possibilities](#future-possibilities)) for the object referenced by `post_state_hash`. In the case depicted, the node which has Block, will download ParentBlock and the referenced GlobalState1, then apply both Deploys and check if the output is the same as the GlobalState2.

## Loose ends

Graph traversal is simple if a known starting point is available, e.g. when beginning from a trusted block that forms a root of the pseudo-tree used in an operation. Some objects are more "dynamic", they are added after the fact and point to existing objects, which must remain immutable.

> * Low levels of this "dynamicism" can be worked around by **introducing refs**. A newly created block has a parent dependency that points in the "wrong direction": From the perspective of a node that has a single trusted block as an ancestor of the new block, there is no connection. By keeping and updating refs like `refs/latest`, we keep multiple handles into the DAG, ensuring we do not miss these.

Author: Let's assume I am starting with a trusted block `T` and download the history, it would look like this

```
                 refs/latest
                   trusted
                      v
T_{-2} <-- T_{-1} <-- T
```

If I now mine a new block `T_{+1}`, the result is

```
                 refs/latest
                   trusted
                      v
T_{-2} <-- T_{-1} <-- T <-- T_{+1} 
```

The block is not "reachable" though, as I have not moved the trusted pointer. I'd suggest to introduce/move `refs/latest` here.

```
                
                  trusted   refs/latest
                      v        v
T_{-2} <-- T_{-1} <-- T <-- T_{+1} 
```

            //       and add tests for known binary representations.
            //       If varint encoding works correctly, this can be a
            //       u32 easily.
pub enum Object {
    Block(Block),
    Deploy(Deloy),
    // ...
}

impl Object {
    /// Deserialize an object, calculating hash while doing so.
    fn deserialize_from<R: Read>(reader: R) -> Result<(Self, ObjectHash), ObjectError> {
        let opts = bincode_opts();
        let mut hr = HashReader512::new(reader);
        let value = opts.deserialize_from(&mut hr)?;

        let (_innerReader, hash) = hr.take();

        Ok((value, hash))
    }

    /// Serialize an object, calculating hash while doing so.
    fn serialize_into<W: Write>(&self, writer: W) -> Result<ObjectHash, ObjectError> {
        let opts = bincode_opts();
        let mut hw = HashWriter512::new(writer);

        opts.serialize_into(hw, &self)?;

        let (_innerWriter, hash) = hw.take();

        Ok(hash)
    }

    /// Calculate hash only. This function may be unnecessary.
    #[inline(always)]
    fn calc_hash(&self) -> Result<ObjectHash, ObjectError> {
        self.serialize(&mut NullWriter)
    }
}

#[derive(Debug)]

> struct ImmutableObject {

Author: My bad. If this CEP gets updated again, I'll be sure to clean up the suggested impl.

    fn object(&self) -> &Object {
        &self.object
    }
}

// not shown: Serialize and Deserialize implementations that forward to the inner
// `Object`'s serialization and deserialization methods.
```

`ImmutableObject` should be the norm for loaded and stored objects, construction of actual `Object`s is likely only going to happen when they are constructed from parts. A `create_unchecked` method can be added to `ImmutableObject` for loading objects from storage, but it is advantageous to hash even loaded objects for extra durability.

_NOTE: the custom iterator made necessary by the design of `iter_outgoing` does not require heap allocations._

```rust
/// An object that supports iteration over its dependencies.
pub trait OutIterable {
    type OutIter : Iterator<Item=ObjectHash>;

    /// Iterate over all hashes referenced by this object directly.
    fn iter_outgoing(self) -> Self::OutIter;
}

enum ObjectOutIter<'a> {
    Block(<Block as OutIterable>::OutIter<'a>),
    Deploy(<Deploy as OutIterable>::OutIter<'a>),
    // ...
}

impl<'a> OutIterable for &'a Object {
    type OutIter = ObjectOutIter<'a>;

    fn iter_outgoing(self) -> Self::OutIter {
        match self {

>             Object::Block(block) => ObjectOutIter::Block(block.iter_outgoing()),

Author: Oof! Point taken.

                BlockHeaderIter::Deploys(block, 0)
            }
            BlockHeaderIter::Deploys(block, idx) if idx < block.deploys.len() => {
                rv = block.deploys[idx].clone();
                BlockHeaderIter::Deploys(block, idx + 1)
            }
            BlockHeaderIter::Deploys(block, idx)
            | BlockHeaderIter::Done => BlockHeaderIter::Done,
        };

        rv
    }
}
```

_NOTE: The above logic is necessary to store and send arbitrary objects using this proposed model; the following logic describes optional convenience methods implemented on `Object` to improve the ergonomics for component developers._

```rust
impl Object {
    fn as_block(&self) -> Option<&Block> {
        match self {
            Object::Block(ref block) => Some(block),
            _ => None
        }
    }

    // ...
}
```

### Storage implementation

Using this proposed model, the storage component decomplects into a single `ObjectStore`, with a simplified set of effects.

```rust
/// Retrieve multiple objects.
// This method always uses the heap, since a single is already quite large.
// Does not return any dependencies, may return less objects than requested.

> pub async fn get_objects(self, object_hash: Vec<ObjectHash>) -> HashMap<ObjectHash, ImmutableObject>;

Author: No, I believe this is a bit of a strawman, sorry!

(Disclaimer: The example code is lagging behind the higher-level CEP)

There would still be a `get_block` method which, like it does now, returns a result of a block - except that the error type now has an "invalid type" variant as well. The `get_object*` methods are a bit more "low-level" for those components that want to deal with objects without caring what is inside.

The "new" approach only gives you more options should you need them in that you can now bulk retrieve objects of multiple types at the same time. It does not take away the old ones.

/// Retrieve multiple objects, including their dependencies.
///
/// If specified, stops at a specific depth. Returns `None` if any requested
/// object was incomplete until `max_depth`.
pub async fn get_objects_with_dependencies(self, object_hash: Vec<ObjectHash>, max_depth: Option<NonZeroUsize>) -> Option<HashMap<ObjectHash, ImmutableObject>>>;

/// Store multiple objects.
///
/// Objects are stored in the reverse order given, so dependencies should be
/// listed behind the actual object.

> pub async fn put_objects(self, objects: Vec<ImmutableObject>);

Author: No, I'd like to avoid this. Having "extra" objects lying around should never be harmful, at worst they will get GC'd at a later point in time. Note that there is no way to remove objects =).

If being able to cancel the put due to some condition, I would assume it's better to batch them or even put them one-by-one.

> pub async fn put_objects(self, objects: Vec<ImmutableObject>);

Reviewer 3: > If being able to cancel the put due to some condition, I would assume it's better to batch them or even put them one-by-one.

And indeed, write operations on the trie do create a changelist of new trie nodes which is only written at the very end.  

On a slightly higher level, currently if we are writing many values to trie at once, they are inserted one by one, but some day we will implement a batched write mechanism - it just hasn't been a priority as of yet.

As I mentioned before, I would be comfortable moving to a transaction-less model so long as there was a good GC story.

```

> The `ObjectStore` continues to use LMDB as a backend, but calculates keys upfront (we might also consider writing to a temporary key and renaming it) and stores the object under a precalculated key. For convenience, given a proper `Hash` trait implementation for `ObjectHash`, a hashmap might be returned.

Author: > Aren't these just the objects' hashes?

Yes, that's what I meant =).

> we might also consider writing to a temporary key and renaming it

If we were to store an object that is brand new and has never had its hash created, it would have to be serialized to do so. So instead of creating another copy in memory, we write it to a temporary key, calculating the hash as we write it. Once that's done, we rename the temporary key to match the hash. All in all it's purely an optimization that saves copying an object on some occasions.

The recursive `_with_dependencies` retrieval functions allows fetching an entire subtree, which is useful when getting a block from storage. For instance, all contained deploys (`max_depth` >= 1) and their Wasm (`max_depth` is >= 2) could be fetched alongside of a block, saving event round trips.

### Execution engine

The execution engine is going to require a change, namely in the hashing function. All values it executes will need to be wrapped in an object enum before being stored/hashed. The exact impact of this is unknown, see the section on [unresolved questions](#unresolved-questions).

Additionally, if Wasm blocks are to be split off, these will need to be fetched separately. If any of these solutions prove to be infeasible, passing around a wrapper around LMDB like we do now is still feasible to keep these operations speedy.

### Gossiping

> Under this proposed model, we no longer require a generic gossiper component; a single statically typed gossiper can handle all objects. The existing functionality can be kept, with some allowed specialization around announcements (such as an incoming deploy received announcement).

Author: As stated above, that needn't be the case in my opinion. It's possible to supply a type-safe interface to consumers!

```rust
/// Gossip a set of objects.
///
/// Dependencies of objects will not be retrieved, checked or gossiped,
/// it is up to the caller to ensure the object itself and its dependencies
/// are in the local storage.
pub async fn gossip_objects(self, objects: Vec<Object>);
```

### Fetching

Under this proposed model, we no longer require a generic fetcher component; a single statically typed fetcher can handle all objects. The existing functionality can be retained with some minor alterations such as the `fetch_object` function described below .

```rust
/// Retrieve an object from storage or the network.
///
/// Retrieves an object
pub async fn fetch_object(self, object_hash: Object,
                                max_depth: Option<NonZeroUsize>,
                                peer: I) -> HashMap<ObjectHash, <FetchResult>>;
```

Here we employ the same `max_depth` limiter as we do with the storage itself.

It is worth noting that given a current block hash `h`, calling `fetch_object(h, None, some_peer)` will eventually fetch the entire state required to join the network!

### Consensus DAG

> There are two ways of embedding consensus DAG data into the system, either the object format accompanies every single type of supported consensus with an enum variant, or a generic variant is used.

Author: The mentioned statement is very specific to our idea of pluggable consensus, there has been some pushback regarding making the actual consensus nodes explicitly known outside of consensus components. That's what those two different alternatives allude to.

For the trie store I would assume it is less controversial due to there only being a fixed set of node types.

> Given that the state trie is essentially a subgraph of the overall block DAG, we'd likely either want to add the Trie type (specialized to Trie<Key, StoredValue>) as variant in Object, or take the trait object approach as below (though I think I prefer the former).

I agree, I also prefer the "closed" approach with Trie variants.

> Indeed, it is a subgraph with very particular structure and growth mechanism - something to keep in mind when considered in the context of OutIterable / ObjectOutIter.

I admit the idea of having a general traversal mechanism might warrant a second look - but there's no rule that says we're only allowed to have one method of traversing the graph =).

> I can't quite tell which approach you favor for the consensus DAG.

Personally, with respect to the things stated above, I like things to be explicit, thus having additional variants in the object graph. Potentially we could have a separate nested enum for these DAG nodes.

The reason I prefer this is that if we ever change the format, we will still have to support shipping these old objects around for quite some time, so they never really go away.

> There are two ways of embedding consensus DAG data into the system, either the object format accompanies every single type of supported consensus with an enum variant, or a generic variant is used.

Reviewer 3: We are in agreement here.

As mentioned, I think the `OutIterable` model can be adapted to state trie traversal without too much fuss.

It is entirely possible to implement pack files as just a serialization of `HashMap<ObjectHash, Object>`.

### Snapshots

A feature currently on the roadmap becomes almost a non-issue with this approach; snapshotting or *checkpointing*, i.e. the ability to start a node without having to download all content from the network, requires no code modification in the sense that if objects are identified by immutable hashes, the local database can simply check if they are present to avoid redownloading them.

Combined with pack files, a snapshot or *checkpoint* of a node is just a specific Block plus all of its dependencies in a single file.

### Finality signatures

[finality-signature]: #finality-signatures

Finality signatures at first glance are an example of a "graph node" that points in the wrong direction: Each signature points to exactly one block, but they will not all be available at the time the block itself is created and its hash finalized. This would usually leave each signature "floating" around, not being a dependency of the block it points to:

!["Floating" finality signatures](../diagrams/0008-finality-sigs-a.png)

> A possible solution to this problem is to define a window `k` during which finality signatures for a block can be attached to its child blocks. As an example, if block `n` is created, finality signatures for this block may be part of any block `(n+1)...(n+k)`.

Reviewer 4: I don't understand why this is a problem in the first place. Why is it so important that everything is a dependency of a single ref?
